### PR TITLE
ASoC: SOF: add DAI_HW_FREE IPC command

### PR DESCRIPTION
**Pull request by brentlu**
This PR is for discussion in https://github.com/thesofproject/sof/pull/4219 to solver the hw_free issue, see comment https://github.com/thesofproject/sof/pull/4219#issuecomment-850195736

There is an STREAM_PCM_FREE IPC command which is sent to a SOF
pipeline in trigger stop. Here we add an new DAI_HW_FREE command which
will be sent to SOF DAI component. The command is sent from BE DAI
link so codec driver has chance to run code in the mute_stream()
callback before the DAI_HW_FREE IPC command is sent.

This is the call flow of pcm playback:

DROP:
    sof's trigger()
        => send STREAM_TRIG_STOP
        => send STREAM_PCM_FREE
HW_FREE:
    sof's hw_free() (FE)
    codec's mute_stream()
    sof's hw_free() (BE)
        => send DAI_HW_FREE
    codec's hw_free()

Signed-off-by: Brent Lu <brent.lu@intel.com>

### CODE REVIEW

File: sound/soc/sof/pcm.c [hunk 1]
```diff
@@ -288,9 +288,41 @@ static int sof_pcm_hw_free(struct snd_soc_component *component,
 	struct snd_sof_pcm *spcm;
 	int ret, err = 0;
 
-	/* nothing to do for BE */
-	if (rtd->dai_link->no_pcm)
+	if (rtd->dai_link->no_pcm) {
```
Comment (lyakh) [inline sound/soc/sof/pcm.c:291]: Interesting, this new IPC message is only needed and only sent in the "no PCM" case, right? Presumably, otherwise the same clean up is triggered in the firmware by an already existing and used IPC? Is a corresponding firmware change already available?

Comment (brentlu) [reply sound/soc/sof/pcm.c:291]: Yes, which is the backend DAI. I forgot to mention the FW counterpart:

https://github.com/thesofproject/sof/pull/4262

File: include/sound/sof/header.h [hunk 2]
```diff
@@ -89,6 +89,7 @@
 /* DAI messages */
 #define SOF_IPC_DAI_CONFIG			SOF_CMD_TYPE(0x001)
 #define SOF_IPC_DAI_LOOPBACK			SOF_CMD_TYPE(0x002)
+#define SOF_IPC_DAI_HW_FREE			SOF_CMD_TYPE(0x004)
```
Comment (bardliao) [inline include/sound/sof/header.h:92]: Maybe rename it to `SOF_IPC_BE_DAI_HW_FREE`. I will assume SOF_IPC_DAI_HW_FREE is for **FE** hw_free.

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (bardliao) [inline sound/soc/sof/pcm.c:291]: It is worthy to have a comment to explain why we need to do this..

**Review by bardliao — COMMENTED** [review]
Thanks @brentlu It is a good idea. However, we will still get problem if codec want to do something in its hw_free() function. I am thinking should we move the task to [sof_pcm_close](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/pcm.c#L582)?

File: include/sound/sof/header.h (hunk 2 above)
Comment (brentlu) [reply include/sound/sof/header.h:92]: how about SOF_IPC_DAI_BE_HW_FREE ?

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (brentlu) [reply sound/soc/sof/pcm.c:291]: hmm...do you think we need a separate function for this?

Comment (brentlu) [reply sound/soc/sof/pcm.c:291]: forget it....the function is in component driver...

Comment (keyonjie) [reply sound/soc/sof/pcm.c:291]: @brentlu if here is late enough? it is in component driver .hw_free() which happen still before dai->driver->ops->hw_free()?

**Comment by plbossart** [comment]
@brentlu it's hard to follow why we need this change. Can elaborate on what the current solution does wrong, and how this is related to the existing changes for the early bitclock enabling?

I am personally completely confused now on how DAIs are handled...

**Comment by uhs987** [comment]
Hi,

1. PCM_FREE is sent in trigger stop instead of hw_free in recent sof linux so we can't rely on it.
2. In pcm.c, we register sof_pcm_hw_free() function to the component driver, so this function will be called twice: once for the FE DAI Link and once for the BE DAI Link. Since ALSA handles FE Link before BE Link for hw_free, we need to send IPC command from the BE component driver's hw_free callback so codec driver has chance to do things before this IPC

FE CPU DAI mute_stream()
FE DAI Link hw_free()
FE component driver hw_free()
=> PCM_FREE could be sent here only if not sent in trigger stop
FE CPU DAI hw_free

BE CPU/CODEC DAI mute_stream()
=> codec driver switching PLL here
BE DAI Link hw_free()
BE component driver hw_free()
=> this PR send HW_FREE here
BE CPU/CODEC DAI hw_free

**Comment by brentlu** [comment]
Hi @plbossart ,

1. PCM_FREE is sent in trigger stop instead of hw_free in recent sof linux so we can't rely on it.
2. In pcm.c, we register sof_pcm_hw_free() function to the component driver, so this function will be called twice: once for the FE DAI Link and once for the BE DAI Link. Since ALSA handles FE Link before BE Link for hw_free, we need to send IPC command from the BE component driver's hw_free callback so codec driver has chance to do things before this IPC

FE CPU DAI mute_stream()
FE DAI Link hw_free()
FE component driver hw_free()
=> PCM_FREE could be sent here only if not sent in trigger stop
FE CPU DAI hw_free

BE CPU/CODEC DAI mute_stream()
=> codec driver switching PLL here
BE DAI Link hw_free()
BE component driver hw_free()
=> this PR send HW_FREE here
BE CPU/CODEC DAI hw_free

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (brentlu) [reply sound/soc/sof/pcm.c:291]: Codecs like cs42l42 switches PLL in the mute_stream() callback so it should be enough. I'm still waiting their validation result.

**Comment by plbossart** [comment]
Sorry @brentlu I don't think this "PCM_FREE is sent in trigger stop instead of hw_free in recent sof linux so we can't rely on it." should be taken as a reference, it makes no sense IMHO. a trigger should not result in a hw_free, that's just wrong. Can you provide a pointer on this so that we all see what you are referring to?

@ranj063 does this ring a bell?

**Comment by brentlu** [comment]
I'm not familiar with the development history of linux side so everything looks like a feature to me...

In the sof_pcm_trigger() function of sound/soc/sof/pcm.c, the reset_hw_params flag is set to TRUE when the trigger command is STOP, then an additional PCM_FREE command will be sent by calling sof_pcm_dsp_pcm_free() function.

static int sof_pcm_trigger(struct snd_soc_component *component,
			   struct snd_pcm_substream *substream, int cmd)
{
...
	switch (cmd) {
...
	case SNDRV_PCM_TRIGGER_STOP:
		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_STOP;
		ipc_first = true;
		reset_hw_params = true;
		break;
...
	/* free PCM if reset_hw_params is set and the STOP IPC is successful */
	if (!ret && reset_hw_params) {
		ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
		if (ret < 0)
			return ret;

**Comment by ranj063** [comment]
> Sorry @brentlu I don't think this "PCM_FREE is sent in trigger stop instead of hw_free in recent sof linux so we can't rely on it." should be taken as a reference, it makes no sense IMHO. a trigger should not result in a hw_free, that's just wrong. Can you provide a pointer on this so that we all see what you are referring to?
> 
> @ranj063 does this ring a bell?

@plbossart yes, thats correct. the PCM_FREE is sent during the stop trigger instead of hw_free and this is done because we dont do hw_free during suspend but just do a stop.

**Comment by plbossart** [comment]
> > Sorry @brentlu I don't think this "PCM_FREE is sent in trigger stop instead of hw_free in recent sof linux so we can't rely on it." should be taken as a reference, it makes no sense IMHO. a trigger should not result in a hw_free, that's just wrong. Can you provide a pointer on this so that we all see what you are referring to?
> > @ranj063 does this ring a bell?
> 
> @plbossart yes, thats correct. the PCM_FREE is sent during the stop trigger instead of hw_free and this is done because we dont do hw_free during suspend but just do a stop.

not following, sorry @ranj063. Also having a hard time figuring out why sof_pcm_dsp_pcm_free() is called from two separate locations.

**Comment by brentlu** [comment]
This is the commit log. Hope it helps.

commit 04c8027764bc82a325d3abc6f39a6a4642a937cb
Author: Kai Vehmanen <kai.vehmanen@linux.intel.com>
Date:   Mon Jul 22 09:13:43 2019 -0500

    ASoC: SOF: reset DMA state in prepare

    When application goes through SUSPEND/STOP->PREPARE->START
    cycle, we should always reprogram the SOF device to start
    DMA from a known state so that hw_ptr/appl_ptrs remain valid.
    This is expected by ALSA core as it resets the buffer
    state as part of prepare (see snd_pcm_do_prepare()).

    Fix the issue by forcing reconfiguration of the FW with
    STREAM_PCM_PARAMS in prepare(). Use combined logic to handle
    prepare and the existing flow to reprogram hw-params after
    system suspend.

    Without the fix, first call to pcm pointer() will return
    an invalid hw_ptr and application may immediately observe XRUN
    status, unless "start_threshold" SW parameter is set to maximum
    value by the application.

    Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>
    Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
    Link: https://lore.kernel.org/r/20190722141402.7194-3-pierre-louis.bossart@linux.intel.com
    Signed-off-by: Mark Brown <broonie@kernel.org>

**Comment by ranj063** [comment]
> not following, sorry @ranj063. Also having a hard time figuring out why sof_pcm_dsp_pcm_free() is called from two separate locations.

@plbossart the rationale is as follows:
1. PCM_FREE IPC sent during trigger stop: During S0ix system suspend, we want to keep the WoV stream alive and free all other streams. If we dont send PCM_FREE during the suspend trigger, restarting streams during resume will fail. This wont be a problem for S3 suspend though because the DSP will be coming out of D3. 
2. PCM_FREE send during sof_pcm_hw_params() is to handle the OSS emulation where we have repeated hw_params without hw_free().

**Comment by plbossart** [comment]
Thanks @ranj063, that helps.

I think there is a conceptual disconnect in the current design of SOF. The usual stimuli (hw_params, prepare, trigger, hw_free) that we would expect from a DAI are handled internally as a consequence of the PCM actions. We typically have a single pipeline at the SOF level, but at the kernel level that same pipeline is viewed as two entities (FE and BE).

I am starting to wonder if we need to have 'component' callbacks which are propagated by firmware, and 'kernel' callbacks that are just a mirror of what a BE can do. Currently we assume that DAIs and DMAs are controlled by the firmware but that's a very strong assumption that isn't true in general.

We could e.g. have DAI IPCs for be_hw_params, be_prepare, be_trigger, be_hw_free that are controlled by the kernel, and can potentially alter the current behavior. For a trigger, it could very well be that the actual link starts only when BOTH the firmware and kernel are ready.

What @brentlu is describing with the hw_free is that we have fused two state machines as a single one, with unintended consequences. I think we could generalize this PR and add new IPCs for all the ALSA stages handled at the BE. It could very well be that these callbacks are not implemented in the majority of the cases, or only result in a log with a timestamp for debug, but that could be useful on non-Intel implementations if e.g. DMAs are handled externally. 

@lgirdwood comments welcome.

In other words, if we are going to add a BE_DAI_HW_FREE, we might as well add all the other stages and handle ONE ABI transition.

**Comment by plbossart** [comment]
I have to back-pedal on my previous comment @brentlu @ranj063 @lgirdwood 

It turns out that we've abused the DAI_CONFIG a great deal, i.e. for HDaudio and SoundWire/ALH we send a DAI_CONFIG IPC during the dai hw_params stage and another DAI_CONFIG IPC during the dai hw_free stages, but in the later case with an invalid channel.

For the SSP we follow the same pattern already but we will send the *same* configuration on hw_params and hw_free.

see for example hda-dai.c

````c
static int ssp_dai_setup_or_free(struct snd_pcm_substream *substream, struct snd_soc_dai *dai,
				 bool setup)
{
	struct snd_soc_component *component;
	struct snd_sof_widget *swidget;
	struct snd_soc_dapm_widget *w;
	struct sof_ipc_fw_version *v;
	struct snd_sof_dev *sdev;

	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		w = dai->playback_widget;
	else
		w = dai->capture_widget;

	swidget = w->dobj.private;
	component = swidget->scomp;
	sdev = snd_soc_component_get_drvdata(component);
	v = &sdev->fw_ready.version;

	/* DAI_CONFIG IPC during hw_params is not supported in older firmware */
	if (v->abi_version < SOF_ABI_VER(3, 18, 0))
		return 0;

	if (setup)
		return hda_ctrl_dai_widget_setup(w);

	return hda_ctrl_dai_widget_free(w);
}

static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
			     struct snd_pcm_hw_params *params,
			     struct snd_soc_dai *dai)
{
	return ssp_dai_setup_or_free(substream, dai, true);
}

static int ssp_dai_hw_free(struct snd_pcm_substream *substream,
			   struct snd_soc_dai *dai)
{
	return ssp_dai_setup_or_free(substream, dai, false);
}

static const struct snd_soc_dai_ops ssp_dai_ops = {
	.hw_params = ssp_dai_hw_params,
	.hw_free = ssp_dai_hw_free,
};
````

and in hda.c

````c
int hda_ctrl_dai_widget_free(struct snd_soc_dapm_widget *w)
{
	struct snd_sof_widget *swidget = w->dobj.private;
	struct snd_soc_component *component = swidget->scomp;
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
	struct sof_ipc_dai_config *config;
	struct snd_sof_dai *sof_dai;
	struct sof_ipc_reply reply;
	int ret;

	sof_dai = swidget->private;

	if (!sof_dai || !sof_dai->dai_config) {
		dev_err(sdev->dev, "error: No config to free DAI %s\n", w->name);
		return -EINVAL;
	}

	/* nothing to do if hw_free() is called without restarting the stream after resume. */
	if (!sof_dai->configured)
		return 0;

	config = &sof_dai->dai_config[sof_dai->current_config];

	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
				  &reply, sizeof(reply));
	if (ret < 0) {
		dev_err(sdev->dev, "error: failed resetting DAI config for %s\n", w->name);
		return ret;
	}

	sof_dai->configured = false;

	return sof_widget_free(sdev, swidget);
}
````

I am not sure if we want to change the existing IPC now, it'd be difficult to split the DAI_CONFIG in two cases in backwards-compatible way. I am also not sure if we want to add a new hw_free callback for the dailink when we have one for the dai already, and for other DAIs we piggy-back on the DAI_CONFIG already.

I think what we are missing in the case of the SSP is an indication that the DAI_CONFIG corresponds to a hw_params (valid flag) or a hw_free case (invalid flag). We can 'easily' add this to the hda-dai-code, and we can use it as information in the firmware to turn-on the clocks or conversely stop them.

**Comment by ranj063** [comment]
> I think what we are missing in the case of the SSP is an indication that the DAI_CONFIG corresponds to a hw_params (valid flag) or a hw_free case (invalid flag). We can 'easily' add this to the hda-dai-code, and we can use it as information in the firmware to turn-on the clocks or conversely stop them.

I agree @plbossart. We can maybe use the early_start bit in the clks_control field in ssp_params() or maybe add a new bit to indicate a late_stop to stop the bclk

**Comment by plbossart** [comment]
@ranj063 @bardliao @brentlu See what I had in mind here:

kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status

I see an underflow so there's probably something missing but the idea of using a flag to indicate hw_params or hw_free is there.

**Comment by ranj063** [comment]
> @ranj063 @bardliao @brentlu See what I had in mind here:
> 
> kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
> SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status
> 
> I see an underflow so there's probably something missing but the idea of using a flag to indicate hw_params or hw_free is there.

LGTM @plbossart

**Comment by bardliao** [comment]
> @ranj063 @bardliao @brentlu See what I had in mind here:
> 
> kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
> SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status
> 
> I see an underflow so there's probably something missing but the idea of using a flag to indicate hw_params or hw_free is there.

It seems that something is required before setting ssp clocks. The clock doesn't enable properly with your branch. [sof_widget_setup](https://github.com/plbossart/sound/blob/fix/ssp-dai-config/sound/soc/sof/intel/hda.c#L67) -> [sof_dai_config_setup](https://github.com/plbossart/sound/blob/fix/ssp-dai-config/sound/soc/sof/sof-audio.c#L46) will send the first `SOF_IPC_DAI_CONFIG` 'IPC and clock will not enable properly ATM. 
Below patch clear config->flags, so ssp clock will be enabled after `sof_widget_setup()`, and it works on my side.
```patch
diff --git a/sound/soc/sof/intel/hda.c b/sound/soc/sof/intel/hda.c
index 2a4c69940b69..62d8690cad0b 100644
--- a/sound/soc/sof/intel/hda.c
+++ b/sound/soc/sof/intel/hda.c
@@ -59,8 +59,6 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w)
        }

        config = &sof_dai->dai_config[sof_dai->current_config];
-       config->flags |= SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
-       config->flags &= ~SOF_DAI_CONFIG_FLAGS_HW_FREE;

        /*
         * For static pipelines, the DAI widget would already be set up and calling
@@ -73,6 +71,9 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w)
                return ret;
        }

+       config->flags |= SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
+       config->flags &= ~SOF_DAI_CONFIG_FLAGS_HW_FREE;
+
        /* send DAI_CONFIG IPC */
        ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
                                  &reply, sizeof(reply));
diff --git a/sound/soc/sof/sof-audio.c b/sound/soc/sof/sof-audio.c
index 5f5c03476932..4fcb03f9036c 100644
--- a/sound/soc/sof/sof-audio.c
+++ b/sound/soc/sof/sof-audio.c
@@ -55,6 +55,8 @@ static int sof_dai_config_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *da
                return -EINVAL;
        }

+       config->flags &= ~SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
+       config->flags &= ~SOF_DAI_CONFIG_FLAGS_HW_FREE;
        ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
                                 &reply, sizeof(reply));
```

**Comment by plbossart** [comment]
Thanks @bardliao, I appreciate the time you've spent debugging my half-baked idea. Testing the changes now.

@ranj063 @juimonen is there a specific reason why we send a DAI_CONFIG IPC from sof_dai_config_setup()?
My traces show tons of calls with the same config, not sure why this is happening. ONE IPC should be enough, no?

also wondering why we send the DAI_CONFIG IPC during the hw_params stage, IIRC on resume the ALSA/ASoC core only use the .prepare stage. We had this issue with SoundWire and the IPC is send in the prepare stage, not sure why other DAIs are different?

**Comment by plbossart** [comment]
@bardliao @brentlu I updated my branches with the fixes suggested by Bard. 

kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status

The results are a bit weird, the bclk starts early and stops late, but if you look at the start, the bclk and fsync don't start at the same time.

start

but they stop at the same time

stop2

with a minor issue that there are a couple of irregular cycles for the bclk

stop1

The stop sequence seems to need more work...

**Comment by plbossart** [comment]
@brentlu @bardliao I updated my branches to reduce the number of configurations and deal with the resume cases.
The only remaining problem I see is the fsync starting late, this isn't expected.

kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status

@brentlu can you double-check how complicated this would be to backport in a Chrome production branch (kernel and firmware?)

**Comment by bardliao** [comment]
> @brentlu @bardliao I updated my branches to reduce the number of configurations and deal with the resume cases.
> The only remaining problem I see is the fsync starting late, this isn't expected.
> 
> kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
> SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status
> 
> @brentlu can you double-check how complicated this would be to backport in a Chrome production branch (kernel and firmware?)

@plbossart I added some comments on your branch. The main question is that is starting clock at prepare stage good enough?

**Comment by brentlu** [comment]
> @brentlu @bardliao I updated my branches to reduce the number of configurations and deal with the resume cases.
> The only remaining problem I see is the fsync starting late, this isn't expected.
> 
> kernel: https://github.com/plbossart/sound/tree/fix/ssp-dai-config
> SOF: https://github.com/plbossart/sof/tree/ssp/dai-config-status
> 
> @brentlu can you double-check how complicated this would be to backport in a Chrome production branch (kernel and firmware?)

I add the comment in https://github.com/thesofproject/sof/pull/4288, I think we can close this PR since we are not adding ipc command for this.
